src/main/java/org/broadinstitute/hellbender/tools/spark/RevertSamSpark.java

```diff
@@ -0,0 +1,706 @@
+                    primaryReads = Collections.singletonList(unpairedRecord);
+                }
+                // Otherwise remove these reads entirely
+                else {
+                    return out.iterator();
+                }
+            }
+
+            // If we've made it this far spit the records into the output!
+            for (final GATKRead rec : primaryReads) {
+                // The only valid quality score encoding scheme is standard; if it's not standard, change it.
+                final FastqQualityFormat recordFormat = readGroupToFormat.get(rec.getReadGroup());
+                if (recordFormat != null && !recordFormat.equals(FastqQualityFormat.Standard)) {
+                    final byte[] quals = rec.getBaseQualities();
+                    for (int i = 0; i < quals.length; i++) {
+                        quals[i] -= SolexaQualityConverter.ILLUMINA_TO_PHRED_SUBTRAHEND;
+                    }
+                    rec.setBaseQualities(quals);
+                }
+                out.add(rec);
+            }
+            return out.iterator();
+        });
+
+    }
+
+    /**
+     * Method which takes an RDD of reads that is guaranteed to have every readname group placed together on the same
+     * partition and maps those so a JavaPairRDD with the readname as the key.
+     */
+    private static JavaPairRDD<String, Iterable<GATKRead>> spanReadsByKey(final JavaRDD<GATKRead> reads) {
+        JavaPairRDD<String, GATKRead> nameReadPairs = reads.mapToPair(read -> new Tuple2<>(read.getName(), read));
+        return SparkUtils.spanByKey(nameReadPairs).flatMapToPair(namedRead -> {
+            // for each name, separate reads by key (group name)
+            List<Tuple2<String, Iterable<GATKRead>>> out = Lists.newArrayList();
+            ListMultimap<String, GATKRead> multi = LinkedListMultimap.create();
+            for (GATKRead read : namedRead._2()) {
+                multi.put(read.getName(), read);
+            }
+            for (String key : multi.keySet()) {
+                // list from Multimap is not serializable by Kryo, so put in a new array list
+                out.add(new Tuple2<>(key, Lists.newArrayList(multi.get(key))));
+            }
+            return out.iterator();
+        });
+    }
+
+    private static JavaRDD<GATKRead> querynameSortReadsIfNecessary(JavaRDD<GATKRead> reads, int numReducers, SAMFileHeader headerForTool) {
+        JavaRDD<GATKRead> sortedReadsForMarking;
+        if (ReadUtils.isReadNameGroupedBam(headerForTool)) {
+            sortedReadsForMarking = reads;
+        } else {
+            headerForTool.setSortOrder(SAMFileHeader.SortOrder.queryname);
+            sortedReadsForMarking = SparkUtils.sortReadsAccordingToHeader(reads, headerForTool, numReducers);
+        }
+        return sortedReadsForMarking;
+    }
+
+
+    private Map<String, SAMFileHeader> getReadgroupHeaderMap(SAMFileHeader inHeader, Map<String, File> writerMap) {
```

Comment:
we're very inconsistent about ReadGroup being one word or two for camelCase purposes.